cpp/src/arrow/compute/test-util.h

```diff
@@ -66,6 +66,44 @@ std::shared_ptr<Array> _MakeArray(const std::shared_ptr<DataType>& type,
   return result;
 }
 
+template <typename Type, typename Enable = void>
+struct DatumEqual {
+  static void EnsureEqual(const Datum& lhs, const Datum& rhs) {}
+};
+
+template <typename Type>
+struct DatumEqual<Type, typename std::enable_if<IsFloatingPoint<Type>::value>::type> {
+  static constexpr double kArbitraryDoubleErrorBound = 1.0;
+  using ScalarType = typename TypeTraits<Type>::ScalarType;
+
+  static void EnsureEqual(const Datum& lhs, const Datum& rhs) {
+    ASSERT_EQ(lhs.kind(), rhs.kind());
+    if (lhs.kind() == Datum::SCALAR) {
```

Comment:
Should assert false or abort on other datum kinds, no?

Reply:
https://issues.apache.org/jira/browse/ARROW-4673

Reply:
I'll also try to include this while implementing ARROW-3123.